Failed to start TiFlash when some CPUs are offline

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

In a machine with 72 logical cpu cores, and let 64 of them off-line.

![6qzOnG0kEG](https://github.com/pingcap/tiflash/assets/60744015/fe570c35-2253-4567-a0c0-816545f1a583)

<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

Start TiFlash successfully.

### 3. What did you see instead (Required)

```
[2024/07/10 14:55:07.141 +08:00] [ERROR] [BaseDaemon.cpp:495] ["Integer divide by zero."] [source=BaseDaemon] [thread_id=2][2024/07/10 14:55:10.890 +08:00] [ERROR] [BaseDaemon.cpp:560] ["\n  0x563cb525a5ce\tfaultSignalHandler(int, siginfo_t*, void*) [tiflash+123495886]\n                \tlibs/libdaemon/src/BaseDaemon.cpp:211\n  0x7fda5033b630\t<unknown symbol> [libpthread.so.0+63024]\n  0x563cafc6e83e\tcomputeAndSetNumberOfPhysicalCPUCores(unsigned short, unsigned short) [tiflash+33400894]\n                \tdbms/src/Common/getNumberOfCPUCores.cpp:66\n  0x563cafc39345\tDB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&) [tiflash+33182533]\n                \tdbms/src/Server/Server.cpp:1077\n  0x563cb6fc7776\tPoco::Util::Application::run() [tiflash+154351478]\n                \tcontrib/poco/Util/src/Application.cpp:335\n  0x563cafc3306f\tDB::Server::run() [tiflash+33157231]\n                \tcontrib/poco/Util/src/ServerApplication.cpp:94\n  0x563cafc4c562\tmainEntryClickHouseServer(int, char**) [tiflash+33260898]\n                \tdbms/src/Server/Server.cpp:1855\n  0x563cafb82010\tmain [tiflash+32432144]\n                \tdbms/src/Server/main.cpp:172\n  0x7fda4faaa555\t__libc_start_main [libc.so.6+140629]"] [source=BaseDaemon] [thread_id=2]
[2024/07/10 14:55:34.790 +08:00] [INFO] [BaseDaemon.cpp:1175] ["Welcome to TiFlash"] [thread_id=1][2024/07/10 14:55:34.790 +08:00] [INFO] [BaseDaemon.cpp:1178] ["TiFlash build info: TiFlash\nRelease Version: v8.2.0-alpha-90-g914aaf71a\nEdition:         Community\nGit Commit Hash: 914aaf71ae9ffae2b9162dbf77c451bf5872ba2f\nGit Branch:      HEAD\nUTC Build Time:  2024-07-10 02:11:01\nEnable Features: jemalloc sm4(GmSSL) avx2 avx512 unwind thinlto\nProfile:         RELWITHDEBINFO\nCompiler:        clang++ 17.0.6\n"] [thread_id=1]
```

### 4. What is your TiFlash version? (Required)

master

--- a/dbms/src/Common/getNumberOfCPUCores.cpp
+++ b/dbms/src/Common/getNumberOfCPUCores.cpp
@@ -46,6 +46,8 @@ void computeAndSetNumberOfPhysicalCPUCores(
     UInt16 number_of_logical_cpu_cores_,
     UInt16 number_of_hardware_physical_cores)
 {
+    number_of_hardware_physical_cores = std::max<UInt16>(1, number_of_hardware_physical_cores);
+
     // First of all, we need to take consideration of two situation:
     //   1. tiflash on physical machine.
     //      In old codes, tiflash needs to set max_threads which is equal to
@@ -62,8 +64,11 @@ void computeAndSetNumberOfPhysicalCPUCores(
     // - `(hardware_logical_cpu_cores / number_of_hardware_physical_cores)` means how many logical cpu core a physical cpu core has.
     // - `number_of_logical_cpu_cores_ / (hardware_logical_cpu_cores / number_of_hardware_physical_cores)` means how many physical cpu cores the tiflash process could use. (Actually, it's needless to get physical cpu cores in virtual environment, but we must ensure the behavior `1` is not broken)
     auto hardware_logical_cpu_cores = std::thread::hardware_concurrency();
-    UInt16 physical_cpu_cores
-        = number_of_logical_cpu_cores_ / (hardware_logical_cpu_cores / number_of_hardware_physical_cores);
+
+    UInt16 cpu_num_per_physical_core = hardware_logical_cpu_cores / number_of_hardware_physical_cores;
+    cpu_num_per_physical_core = std::max<UInt16>(1, cpu_num_per_physical_core);
+
+    UInt16 physical_cpu_cores = number_of_logical_cpu_cores_ / cpu_num_per_physical_core;
     CPUCores::number_of_physical_cpu_cores = physical_cpu_cores > 0 ? physical_cpu_cores : 1;
     LOG_INFO(
         DB::Logger::get(),

--- a/dbms/src/Common/getNumberOfCPUCores.cpp
+++ b/dbms/src/Common/getNumberOfCPUCores.cpp
@@ -65,10 +65,10 @@ void computeAndSetNumberOfPhysicalCPUCores(
     // - `number_of_logical_cpu_cores_ / (hardware_logical_cpu_cores / number_of_hardware_physical_cores)` means how many physical cpu cores the tiflash process could use. (Actually, it's needless to get physical cpu cores in virtual environment, but we must ensure the behavior `1` is not broken)
     auto hardware_logical_cpu_cores = std::thread::hardware_concurrency();
 
-    UInt16 cpu_num_per_physical_core = hardware_logical_cpu_cores / number_of_hardware_physical_cores;
-    cpu_num_per_physical_core = std::max<UInt16>(1, cpu_num_per_physical_core);
+    UInt16 thread_num_per_physical_core = hardware_logical_cpu_cores / number_of_hardware_physical_cores;
+    thread_num_per_physical_core = std::max<UInt16>(1, thread_num_per_physical_core);
 
-    UInt16 physical_cpu_cores = number_of_logical_cpu_cores_ / cpu_num_per_physical_core;
+    UInt16 physical_cpu_cores = number_of_logical_cpu_cores_ / thread_num_per_physical_core;
     CPUCores::number_of_physical_cpu_cores = physical_cpu_cores > 0 ? physical_cpu_cores : 1;
     LOG_INFO(
         DB::Logger::get(),

--- a/dbms/src/Common/tests/gtest_miscellaneous.cpp
+++ b/dbms/src/Common/tests/gtest_miscellaneous.cpp
@@ -1,0 +1,46 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <TestUtils/TiFlashTestBasic.h>
+#include <Common/getNumberOfCPUCores.h>
+
+namespace DB
+{
+namespace tests
+{
+
+TEST(CommonMiscellaneousTest, cpuRelated)
+{
+    auto hardware_logical_cpu_cores = std::thread::hardware_concurrency();
+    UInt16 number_of_logical_cpu_cores = std::thread::hardware_concurrency();
+
+    computeAndSetNumberOfPhysicalCPUCores(number_of_logical_cpu_cores, hardware_logical_cpu_cores);
+    ASSERT_EQ(getNumberOfPhysicalCPUCores(), hardware_logical_cpu_cores);
+    
+    computeAndSetNumberOfPhysicalCPUCores(number_of_logical_cpu_cores, 0);
+    ASSERT_EQ(1, getNumberOfPhysicalCPUCores());
+
+    UInt16 test_num = 123;
+    computeAndSetNumberOfPhysicalCPUCores(test_num, hardware_logical_cpu_cores * 2);
+    ASSERT_EQ(test_num, getNumberOfPhysicalCPUCores());
+
+    computeAndSetNumberOfPhysicalCPUCores(1, hardware_logical_cpu_cores * 2);
+    ASSERT_EQ(1, getNumberOfPhysicalCPUCores());
+
+    computeAndSetNumberOfPhysicalCPUCores(0, hardware_logical_cpu_cores);
+    ASSERT_EQ(1, getNumberOfPhysicalCPUCores());
+}
+
+} // namespace tests
+} // namespace DB

--- a/dbms/src/Common/tests/gtest_miscellaneous.cpp
+++ b/dbms/src/Common/tests/gtest_miscellaneous.cpp
@@ -12,8 +12,8 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-#include <TestUtils/TiFlashTestBasic.h>
 #include <Common/getNumberOfCPUCores.h>
+#include <TestUtils/TiFlashTestBasic.h>
 
 namespace DB
 {
@@ -27,7 +27,7 @@ TEST(CommonMiscellaneousTest, cpuRelated)
 
     computeAndSetNumberOfPhysicalCPUCores(number_of_logical_cpu_cores, hardware_logical_cpu_cores);
     ASSERT_EQ(getNumberOfPhysicalCPUCores(), hardware_logical_cpu_cores);
-    
+
     computeAndSetNumberOfPhysicalCPUCores(number_of_logical_cpu_cores, 0);
     ASSERT_EQ(1, getNumberOfPhysicalCPUCores());
 

